### ISSUE TITLE
Cellular-context baseReg/baseMLP train perturbation response in the reverse direction

### ISSUE THREAD

**Issue by dnncha** [bm2-lab/scPerturBench#13]
## Summary

I found a reproducible direction mismatch in the released cellular-context baselines. In both the i.i.d. and o.o.d. `baseReg.py` and `baseMLP.py` implementations, `generatePairedSample(...)` returns **perturbed cells first and control cells second**, while the caller assigns those outputs to `Xtr, ytr` and later predicts from **control** cells to compare against **stimulated** cells.

That means these baselines are trained as:

`perturbed -> control`

but used/evaluated as if they had learned:

`control -> perturbed`.

This report is limited to the released baseline implementations. It does **not** claim that the paper's overall conclusions are invalid, and I have not yet rerun the publication datasets or recomputed the leaderboard.

## Affected released source

Pinned repository commit inspected: `6e24e7a9827e55d4567d2139427be9af0d1e7a6c`.

Affected files:

- `Cellular_context_generalization/o.o.d./baseReg.py`
- `Cellular_context_generalization/i.i.d./baseReg.py`
- `Cellular_context_generalization/o.o.d./baseMLP.py`
- `Cellular_context_generalization/i.i.d./baseMLP.py`

All four have the same pattern:

```python
# helper
return annList_Pertb, annList_control

# caller
Xtr_anndata, ytr_anndata = generatePairedSample(...)
Xtr = Xtr_anndata.X
ytr = ytr_anndata.X
...
Xte_anndata = ... control ...
...
ypred = model(... Xte ...)
...
yte_anndata = ... perturbed/stimulated ...
```

For the o.o.d. files, `outSample` is excluded from training as intended; the issue is the direction of the paired mapping.

## Minimal correction

The minimal correction in each of the four helpers is:

```diff
-    return annList_Pertb, annList_control
+    return annList_control, annList_Pertb
```

No model hyperparameters, cell-selection rules, test-set definitions, output labels, or evaluation metrics need to change for this correction.

## Executable reproduction

I validated the actual released source functions on controlled synthetic expression data using the real scikit-learn PCA/Ridge path for the linear baseline, with in-memory adapters only for AnnData/scanpy/filesystem boundaries.

The synthetic perturbation was constructed so the true average perturbation response is approximately `+2` expression units.

Results:

| Setting | True mean response | Released source | Corrected source |
|---|---:|---:|---:|
| cellular-context o.o.d. | +2.00000 | about **-1.99973** | about **+2.00005** |
| cellular-context i.i.d. | +2.00000 | about **-1.99981** | about **+1.99999** |

So in this fixture the released baseline learns almost exactly the inverse perturbation response; changing only the helper return order restores the intended direction.

I also exercised positive, negative, and zero perturbation effects across multiple seeds and both evaluation settings. Zero-effect controls remain approximately zero.

Focused regression result across the four released source files:

- released source: **26 failed, 26 passed**
- four-line corrected source: **52 passed**

For the MLP code, the focused tests execute the data-selection/caller path up to the actual `trainModel` call and verify the tensors are presented in the intended direction. They do **not** perform GPU training, so I am not claiming a full native neural rerun.

## Scientific scope / impact not yet established

This demonstrates an implementation-level correctness defect in two released cellular-context baselines. It does **not** by itself establish how much any published metric or ranking changes.

The next decisive check is a native rerun of the affected publication analyses using fresh output paths, first reproducing the released baseline results and then changing only the return order. The i.i.d. scripts skip existing prediction files, so corrected reruns should avoid reusing old outputs.

If these exact scripts (or equivalent versions) generated the paper's reported baseline values, then the affected baseline metrics should be recomputed. If a different unpublished/private implementation generated the paper results, provenance of that implementation would resolve the publication-level impact.

Associated paper: *Benchmarking algorithms for generalizable single-cell perturbation response prediction*, Nature Methods, DOI `10.1038/s41592-025-02980-0`.

## Prior-report check

I checked the repository's existing issues/comments and did not find this exact baseline training-direction reversal reported. I did find the separate CPA counterfactual-evaluation issue already discussed in #1; this report is independent of that known issue.

## Suggested next step

I can provide the complete focused regression harness / patch and help rerun the affected native analyses. The smallest production change is the four return-order edits above.

Prepared with AI-assisted code review and executable before/after tests. No allegation of misconduct is intended.

**Comment by weizhiting** [comment]
Dear Donncha O’Toole,

Thank you very much for your careful report and for bringing this issue to our attention.

We have checked the released GitHub implementation and confirm your finding. The reported cellular-context baseline results were generated from the released code, and, as you pointed out, `generatePairedSample(...)` returns the perturbed cells and control cells in the reversed order. This means that the cellular-context `baseReg` and `baseMLP` baseline results are affected by this implementation error.

We have preserved the original released source under `v1.0-paper-release` for provenance and reproducibility, and have released the minimal correction under `v1.0.1-baseline-direction-fix`. The correction is limited to the four return-order edits you identified; no model hyperparameters, cell-selection rules, test definitions, output labels, or evaluation metrics were changed.

We are now rerunning the affected analyses using the corrected code and fresh output paths. At this stage, we are not making a claim about the magnitude of the effect on the published metrics or rankings until the native reruns are complete. Because I have since moved from my previous institution to a new institution, I need to re-download and reconfigure the relevant datasets and computational environment. Therefore, the full rerun and updated results may take some time. Once the rerun is complete, we will update this issue and report the corrected baseline results and any observed changes to the affected metrics.

Thank you again for identifying this issue and for the constructive way in which you reported it. We appreciate your help.

Best regards,
Zhiting Wei
